LSP

```
10�� 31, 2022 5:56:17 ���� org.javacs.lsp.LSP connect
����: Unexpected IllegalAccessException occurred (Gson 2.9.1). Certain ReflectionAccessFilter features require Java >= 9 to work correctly. If you are not using ReflectionAccessFilter, report this to the Gson maintainers.
java.lang.RuntimeException: Unexpected IllegalAccessException occurred (Gson 2.9.1). Certain ReflectionAccessFilter features require Java >= 9 to work correctly. If you are not using ReflectionAccessFilter, report this to the Gson maintainers.
	at com.google.gson.internal.reflect.ReflectionHelper.createExceptionForUnexpectedIllegalAccess(ReflectionHelper.java:69)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:272)
	at com.google.gson.Gson.fromJson(Gson.java:1058)
	at com.google.gson.Gson.fromJson(Gson.java:1129)
	at com.google.gson.Gson.fromJson(Gson.java:1101)
	at org.javacs.lsp.LSP.connect(LSP.java:242)
	at org.aya.lsp.LspMain.call(LspMain.java:35)
	at org.aya.lsp.LspMain.call(LspMain.java:22)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at org.aya.lsp.LspMain.main(LspMain.java:24)
```

Repro: Ctrl+L Ctrl+P

```
Caused by: java.lang.IllegalAccessException: Can not set final java.net.URI field org.aya.lsp.models.ComputeTermResult$Params.uri to java.net.URI
	at java.base/jdk.internal.reflect.FieldAccessorImpl.throwFinalFieldIllegalAccessException(FieldAccessorImpl.java:137)
	at java.base/jdk.internal.reflect.FieldAccessorImpl.throwFinalFieldIllegalAccessException(FieldAccessorImpl.java:141)
	at java.base/jdk.internal.reflect.MethodHandleObjectFieldAccessorImpl.set(MethodHandleObjectFieldAccessorImpl.java:103)
	at java.base/java.lang.reflect.Field.set(Field.java:820)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:166)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:266)
```

--- a/LICENSE
+++ b/LICENSE
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2021 Yinsen (Tesla) Zhang
+Copyright (c) 2021-2022 Tesla (Yinsen) Zhang
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal

--- a/base/src/main/java/org/aya/concrete/Expr.java
+++ b/base/src/main/java/org/aya/concrete/Expr.java
@@ -54,12 +54,12 @@ public sealed interface Expr extends AyaDocile, SourceNode, Restr.TermLike<Expr>
         var param = pi.param().descent(f);
         var last = f.apply(pi.last());
         if (param == pi.param() && last == pi.last()) yield pi;
-        yield new Expr.PiExpr(pi.sourcePos(), pi.co(), param, last);
+        yield new Expr.PiExpr(pi.sourcePos(), param, last);
       }
       case Expr.SigmaExpr sigma -> {
         var params = sigma.params().map(param -> param.descent(f));
         if (params.sameElements(sigma.params(), true)) yield sigma;
-        yield new Expr.SigmaExpr(sigma.sourcePos(), sigma.co(), params);
+        yield new Expr.SigmaExpr(sigma.sourcePos(), params);
       }
       case Expr.RawSortExpr rawType -> rawType;
       case Expr.LiftExpr lift -> {
@@ -280,7 +280,6 @@ public sealed interface Expr extends AyaDocile, SourceNode, Restr.TermLike<Expr>
    */
   record PiExpr(
     @NotNull SourcePos sourcePos,
-    boolean co,
     @NotNull Expr.Param param,
     @NotNull Expr last
   ) implements Expr {}
@@ -339,7 +338,6 @@ public sealed interface Expr extends AyaDocile, SourceNode, Restr.TermLike<Expr>
    */
   record SigmaExpr(
     @NotNull SourcePos sourcePos,
-    boolean co,
     @NotNull ImmutableSeq<@NotNull Param> params
   ) implements Expr {}
 

--- a/base/src/main/java/org/aya/resolve/visitor/ExprResolver.java
+++ b/base/src/main/java/org/aya/resolve/visitor/ExprResolver.java
@@ -110,14 +110,14 @@ public record ExprResolver(
         yield new Expr.HoleExpr(hole.sourcePos(), hole.explicit(), h, hole.accessibleLocal());
       }
       case Expr.PartEl el -> partial(ctx, el);
-      case Expr.Path path -> {
-        var newCtx = resolveCubeParams(path.params(), ctx);
-        var par = partial(newCtx, path.partial());
-        var type = resolve(path.type(), newCtx);
-        if (type == path.type() && par == path.partial()) yield path;
-        yield new Expr.Path(path.sourcePos(), path.params(), type, par);
+      case Expr.Path(var pos, var params, var ty, var partial) -> {
+        var newCtx = resolveCubeParams(params, ctx);
+        var par = partial(newCtx, partial);
+        var type = resolve(ty, newCtx);
+        if (type == ty && par == partial) yield expr;
+        yield new Expr.Path(pos, params, type, par);
       }
-      case Expr.Array arrayExpr -> arrayExpr.arrayBlock().fold(
+      case Expr.Array(var pos, var array) -> array.fold(
         left -> {
           var bindName = resolve(left.bindName(), ctx);
           var pureName = resolve(left.pureName(), ctx);
@@ -125,24 +125,23 @@ public record ExprResolver(
           var generator = resolve(left.generator(), bindsCtx._2);
 
           if (generator == left.generator() && bindsCtx._1.sameElements(left.binds()) && bindName == left.bindName() && pureName == left.pureName()) {
-            return arrayExpr;
+            return expr;
           } else {
-            return Expr.Array.newGenerator(arrayExpr.sourcePos(), generator, bindsCtx._1, bindName, pureName);
+            return Expr.Array.newGenerator(pos, generator, bindsCtx._1, bindName, pureName);
           }
         },
         right -> {
           var exprs = right.exprList().map(e -> resolve(e, ctx));
 
           if (exprs.sameElements(right.exprList())) {
-            return arrayExpr;
+            return expr;
           } else {
-            return Expr.Array.newList(arrayExpr.sourcePos(), exprs);
+            return Expr.Array.newList(pos, exprs);
           }
         }
       );
-      case Expr.UnresolvedExpr unresolved -> {
-        var sourcePos = unresolved.sourcePos();
-        yield switch (ctx.get(unresolved.name())) {
+      case Expr.UnresolvedExpr(var pos, var name) -> {
+        yield switch (ctx.get(name)) {
           case GeneralizedVar generalized -> {
             if (options.allowGeneralized) {
               // Ordered set semantics. Do not expect too many generalized vars.
@@ -153,8 +152,8 @@ public record ExprResolver(
                 addReference(owner);
               }
             } else if (!allowedGeneralizes.containsKey(generalized))
-              generalizedUnavailable(ctx, sourcePos, generalized);
-            yield new Expr.RefExpr(sourcePos, allowedGeneralizes.get(generalized).ref());
+              generalizedUnavailable(ctx, pos, generalized);
+            yield new Expr.RefExpr(pos, allowedGeneralizes.get(generalized).ref());
           }
           case DefVar<?, ?> ref -> {
             switch (ref.concrete) {
@@ -165,15 +164,15 @@ public record ExprResolver(
               }
               case TyckUnit unit -> addReference(unit);
             }
-            yield new Expr.RefExpr(sourcePos, ref);
+            yield new Expr.RefExpr(pos, ref);
           }
-          case AnyVar var -> new Expr.RefExpr(sourcePos, var);
+          case AnyVar var -> new Expr.RefExpr(pos, var);
         };
       }
-      case Expr.Idiom idiom -> {
-        var newNames = idiom.names().fmap(e -> resolve(e, ctx));
-        var newBody = idiom.barredApps().map(e -> resolve(e, ctx));
-        yield new Expr.Idiom(idiom.sourcePos(), newNames, newBody);
+      case Expr.Idiom(var pos, var names, var barred) -> {
+        var newNames = names.fmap(e -> resolve(e, ctx));
+        var newBody = barred.map(e -> resolve(e, ctx));
+        yield new Expr.Idiom(pos, newNames, newBody);
       }
       default -> expr;
     };

--- a/base/src/main/java/org/aya/resolve/visitor/ExprResolver.java
+++ b/base/src/main/java/org/aya/resolve/visitor/ExprResolver.java
@@ -96,12 +96,12 @@ public record ExprResolver(
       }
       case Expr.SigmaExpr sigma -> {
         var params = resolveParams(sigma.params(), ctx);
-        yield new Expr.SigmaExpr(sigma.sourcePos(), sigma.co(), params._1.toImmutableSeq());
+        yield new Expr.SigmaExpr(sigma.sourcePos(), params._1.toImmutableSeq());
       }
       case Expr.PiExpr pi -> {
         var param = resolveParam(pi.param(), ctx);
         var last = resolve(pi.last(), param._2);
-        yield new Expr.PiExpr(pi.sourcePos(), pi.co(), param._1, last);
+        yield new Expr.PiExpr(pi.sourcePos(), param._1, last);
       }
       case Expr.HoleExpr hole -> {
         hole.accessibleLocal().set(ctx.collect(MutableList.create()).toImmutableSeq());

--- a/base/src/main/java/org/aya/resolve/visitor/ExprResolver.java
+++ b/base/src/main/java/org/aya/resolve/visitor/ExprResolver.java
@@ -140,35 +140,33 @@ public record ExprResolver(
           }
         }
       );
-      case Expr.UnresolvedExpr(var pos, var name) -> {
-        yield switch (ctx.get(name)) {
-          case GeneralizedVar generalized -> {
-            if (options.allowGeneralized) {
-              // Ordered set semantics. Do not expect too many generalized vars.
-              if (!allowedGeneralizes.containsKey(generalized)) {
-                var owner = generalized.owner;
-                assert owner != null : "Sanity check";
-                allowedGeneralizes.put(generalized, owner.toExpr(false, generalized.toLocal()));
-                addReference(owner);
-              }
-            } else if (!allowedGeneralizes.containsKey(generalized))
-              generalizedUnavailable(ctx, pos, generalized);
-            yield new Expr.RefExpr(pos, allowedGeneralizes.get(generalized).ref());
-          }
-          case DefVar<?, ?> ref -> {
-            switch (ref.concrete) {
-              case null -> {
-                // RefExpr is referring to a serialized core which is already tycked.
-                // Collecting tyck order for tycked terms is unnecessary, just skip.
-                assert ref.core != null; // ensure it is tycked
-              }
-              case TyckUnit unit -> addReference(unit);
+      case Expr.UnresolvedExpr(var pos, var name) -> switch (ctx.get(name)) {
+        case GeneralizedVar generalized -> {
+          if (options.allowGeneralized) {
+            // Ordered set semantics. Do not expect too many generalized vars.
+            if (!allowedGeneralizes.containsKey(generalized)) {
+              var owner = generalized.owner;
+              assert owner != null : "Sanity check";
+              allowedGeneralizes.put(generalized, owner.toExpr(false, generalized.toLocal()));
+              addReference(owner);
             }
-            yield new Expr.RefExpr(pos, ref);
+          } else if (!allowedGeneralizes.containsKey(generalized))
+            generalizedUnavailable(ctx, pos, generalized);
+          yield new Expr.RefExpr(pos, allowedGeneralizes.get(generalized).ref());
+        }
+        case DefVar<?, ?> ref -> {
+          switch (ref.concrete) {
+            case null -> {
+              // RefExpr is referring to a serialized core which is already tycked.
+              // Collecting tyck order for tycked terms is unnecessary, just skip.
+              assert ref.core != null; // ensure it is tycked
+            }
+            case TyckUnit unit -> addReference(unit);
           }
-          case AnyVar var -> new Expr.RefExpr(pos, var);
-        };
-      }
+          yield new Expr.RefExpr(pos, ref);
+        }
+        case AnyVar var -> new Expr.RefExpr(pos, var);
+      };
       case Expr.Idiom(var pos, var names, var barred) -> {
         var newNames = names.fmap(e -> resolve(e, ctx));
         var newBody = barred.map(e -> resolve(e, ctx));

--- a/base/src/test/java/org/aya/concrete/ParseTest.java
+++ b/base/src/test/java/org/aya/concrete/ParseTest.java
@@ -123,14 +123,14 @@ public class ParseTest {
     assertTrue(parseExpr("λ a => a") instanceof Expr.LamExpr);
     assertTrue(parseExpr("\\ a => a") instanceof Expr.LamExpr);
     assertTrue(parseExpr("\\ a b => a") instanceof Expr.LamExpr);
-    assertTrue(parseExpr("Π a -> a") instanceof Expr.PiExpr dt && !dt.co());
-    assertTrue(parseExpr("Pi a -> a") instanceof Expr.PiExpr dt && !dt.co());
+    assertTrue(parseExpr("Π a -> a") instanceof Expr.PiExpr dt);
+    assertTrue(parseExpr("Pi a -> a") instanceof Expr.PiExpr dt);
     assertTrue(parseExpr("Pi a b -> a") instanceof Expr.PiExpr dt
-      && !dt.co() && dt.last() instanceof Expr.PiExpr);
-    assertTrue(parseExpr("Σ a ** b") instanceof Expr.SigmaExpr dt && !dt.co());
-    assertTrue(parseExpr("Sig a ** b") instanceof Expr.SigmaExpr dt && !dt.co());
-    assertTrue(parseExpr("Sig a b ** c") instanceof Expr.SigmaExpr dt && !dt.co());
-    assertTrue(parseExpr("Pi (x : Sig a ** b) -> c") instanceof Expr.PiExpr dt && !dt.co() && dt.param().type() instanceof Expr.SigmaExpr);
+      && dt.last() instanceof Expr.PiExpr);
+    assertTrue(parseExpr("Σ a ** b") instanceof Expr.SigmaExpr dt);
+    assertTrue(parseExpr("Sig a ** b") instanceof Expr.SigmaExpr dt);
+    assertTrue(parseExpr("Sig a b ** c") instanceof Expr.SigmaExpr dt);
+    assertTrue(parseExpr("Pi (x : Sig a ** b) -> c") instanceof Expr.PiExpr dt && dt.param().type() instanceof Expr.SigmaExpr);
     parseTo("(f a) . 1", new Expr.ProjExpr(
       SourcePos.NONE,
       new Expr.BinOpSeq(

--- a/cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java
+++ b/cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java
@@ -528,7 +528,7 @@ public record AyaGKProducer(
       var to = expr(exprs.get(1));
       var paramPos = sourcePosOf(expr0);
       var param = new Expr.Param(paramPos, Constants.randomlyNamed(paramPos), expr(expr0), true);
-      return new Expr.PiExpr(pos, false, param, to);
+      return new Expr.PiExpr(pos, param, to);
     }
     if (node.is(NEW_EXPR)) {
       var struct = expr(node.child(EXPR));
@@ -553,7 +553,7 @@ public record AyaGKProducer(
       expr(node.child(EXPR)));
     if (node.is(SIGMA_EXPR)) {
       var last = expr(node.child(EXPR));
-      return new Expr.SigmaExpr(pos, false,
+      return new Expr.SigmaExpr(pos,
         telescope(node.childrenOfType(TELE).map(x -> x))
           .appended(new Expr.Param(last.sourcePos(), LocalVar.IGNORED, last, true)));
     }
@@ -678,7 +678,7 @@ public record AyaGKProducer(
     if (params.isEmpty()) return body;
     var drop = params.drop(1);
     return new Expr.PiExpr(
-      sourcePos, co, params.first(),
+      sourcePos, params.first(),
       buildPi(body.sourcePos().sourcePosForSubExpr(sourcePos.file(),
         drop.map(Expr.Param::sourcePos)), co, drop, body));
   }

--- a/lsp/src/main/java/org/aya/lsp/actions/ComputeTerm.java
+++ b/lsp/src/main/java/org/aya/lsp/actions/ComputeTerm.java
@@ -1,4 +1,4 @@
-// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Copyright (c) 2020-2022 Tesla (Yinsen) Zhang.
 // Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
 package org.aya.lsp.actions;
 
@@ -37,7 +37,7 @@ public final class ComputeTerm implements SyntaxNodeAction.Cursor {
     }
 
     public static @NotNull Kind whnf() {
-      return new Kind((fac, term)  -> term.wellTyped().normalize(new TyckState(fac), NormalizeMode.WHNF));
+      return new Kind((fac, term) -> term.wellTyped().normalize(new TyckState(fac), NormalizeMode.WHNF));
     }
   }
 
@@ -50,7 +50,7 @@ public final class ComputeTerm implements SyntaxNodeAction.Cursor {
   public @NotNull ComputeTermResult invoke(ComputeTermResult.Params params) {
     var program = source.program().get();
     if (program == null) return ComputeTermResult.bad(params);
-    visitAll(program, new XY(params.position()));
+    visitAll(program, new XY(params.position));
     return result == null ? ComputeTermResult.bad(params) : ComputeTermResult.good(params, result);
   }
 

--- a/lsp/src/main/java/org/aya/lsp/models/ComputeTermResult.java
+++ b/lsp/src/main/java/org/aya/lsp/models/ComputeTermResult.java
@@ -14,11 +14,12 @@ import org.jetbrains.annotations.Nullable;
 import java.net.URI;
 
 /**
- * @param uri the uri to the file
  * @param computed null if failed
  */
 public record ComputeTermResult(@NotNull URI uri, @Nullable String computed, @NotNull Range range) {
-  public record Params(@NotNull URI uri, @NotNull Position position) {
+  public static class Params {
+    public URI uri;
+    public Position position;
   }
 
   public static @NotNull ComputeTermResult bad(@NotNull Params params) {

--- a/lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java
+++ b/lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java
@@ -367,7 +367,7 @@ public class AyaLanguageServer implements LanguageServer {
   }
 
   public ComputeTermResult computeTerm(@NotNull ComputeTermResult.Params input, ComputeTerm.Kind type) {
-    var source = find(input.uri());
+    var source = find(input.uri);
     if (source == null) return ComputeTermResult.bad(input);
     return new ComputeTerm(source, type, primFactory(source.owner())).invoke(input);
   }

